pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [ALTER PK failed validation: expected 0 rows, found 4]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20577856?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20577856?buildTab=artifacts#/) on master @ [bb1c098ecc7058a2b24a23820c5a8e235dfa714c](https://github.com/cockroachdb/cockroach/commits/bb1c098ecc7058a2b24a23820c5a8e235dfa714c):


```
     true,
     true,
     true,
     true,
     true,
     true,
     true,
     false
    ]
   }
  ],
  "previousStatements": [
   "ALTER TABLE public.table_w0_141 ALTER PRIMARY KEY USING COLUMNS (col141_w0_147, col141_w0_149, col141_w0_152) USING HASH"
  ]
 }
}
    schema_change_external_test.go:165: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:165
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (XX000): executing declarative schema change PostCommitPhase stage 7 of 16 with 3 ValidationType ops (rollback=false) for ALTER TABLE: validation of index crdb_internal_index_4_name_placeholder failed: expected 0 rows, found 4 (SQLSTATE XX000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:682
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:154
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1223
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (XX000): executing declarative schema change PostCommitPhase stage 7 of 16 with 3 ValidationType ops (rollback=false) for ALTER TABLE: validation of index crdb_internal_index_4_name_placeholder failed: expected 0 rows, found 4 (SQLSTATE XX000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 84
Total Schema Statements Succeeded = 84
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 57
Total Transactions Rolled Back = 40
Total Transactions Executed = 97
    schema_change_external_test.go:118: backup, tracing data, and system table dumps in /artifacts/tmp/_tmp/67537770b84c3a4ea177539c1e46bd43/logTestWorkload4294111833
    schema_change_external_test.go:127: running DROP with use_declarative_schema_changer = on
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/67537770b84c3a4ea177539c1e46bd43/logTestWorkload4294111833
--- FAIL: TestWorkload (63.14s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #154715 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [insert violates foreign key constraint] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3.21-rc]
- #154221 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [depends-on relation is dropped during DROP SCHEMA CASCADE] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.3]
- #153823 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [did not received expected DependentObjectsStillExist error] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3]
- #153126 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [column &#34;crdb_internal_column_3_name_placeholder&#34; does not exist] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3.20-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55258

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -1785,8 +1785,10 @@ func countExpectedRowsForInvertedIndex(
 	if withFirstMutationPublic {
 		// Make the mutations public in an in-memory copy of the descriptor and
 		// add it to the Collection's synthetic descriptors, so that we can use
-		// SQL below to perform the validation.
-		fakeDesc, err := tableDesc.MakeFirstMutationPublic(catalog.IgnoreConstraints, catalog.RetainDroppingColumns)
+		// SQL below to perform the validation. Avoid making PK swaps public, otherwise
+		// in the declarative schema changer we can select incomplete primary indexes
+		// for validation.
+		fakeDesc, err := tableDesc.MakeFirstMutationPublic(catalog.IgnoreConstraints, catalog.RetainDroppingColumns, catalog.IgnorePKSwaps)
 		if err != nil {
 			return 0, err
 		}

--- a/pkg/sql/logictest/testdata/logic_test/alter_table
+++ b/pkg/sql/logictest/testdata/logic_test/alter_table
@@ -5265,3 +5265,23 @@ statement ok
 RESET experimental_enable_unique_without_index_constraints
 
 subtest end
+
+# Tests for a regression found by #155145 where the wrong primary index could
+# be scanned for when validating inverted indexes recreated after an ALTER
+# PRIMARY KEY.
+subtest validate_alter_pk_with_inverted
+
+statement ok
+CREATE TABLE t_alter_pk_with_inverted(
+    a STRING NOT NULL,
+    d DECIMAL[],
+    INVERTED INDEX (d)
+  );
+
+statement ok
+INSERT INTO t_alter_pk_with_inverted VALUES ('test', ARRAY[1.0, 2.0]);
+
+statement ok
+ALTER TABLE t_alter_pk_with_inverted ALTER PRIMARY KEY USING COLUMNS (a) USING HASH;
+
+subtest end

